### PR TITLE
Add basic authenticator and file based userstore API

### PR DESCRIPTION
**Pull request by isurulucky**
## Purpose

> Provides a way to secure Services with basic authentication, with a simple file based userstore.

### CODE REVIEW

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/security/AuthorizationTest.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+public class AuthorizationTest {
+    private static final Log log = LogFactory.getLog(AuthorizationTest.class);
+    private CompileResult compileResult;
+
+    private final String ballerinaConf = "ballerina.conf";
+    private ConfigRegistry registry;
+    private final String userDir = "user.dir";
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/security/AuthorizationTest.java]: Above 2 configurations can be converted to local variables

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/security/BasicAuthenticationTest.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+public class BasicAuthenticationTest {
+    private static final Log log = LogFactory.getLog(BasicAuthenticationTest.class);
+    private CompileResult compileResult;
+
+    private final String ballerinaConf = "ballerina.conf";
+    private ConfigRegistry registry;
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/security/BasicAuthenticationTest.java]: Shall we make above variables local?

File: stdlib/ballerina-builtin/src/main/ballerina/ballerina/security/authentication/basic/authn-interceptor.bal [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        authenticator = createAuthenticator({}, utils:createCache(AUTH_CACHE));
+    }
+    AuthenticationInfo authInfo;
+    // check in the cache - cache key is the sha256 hash of the basic auth header value
+    string basicAuthCacheKey = crypto:getHash(basicAuthHeaderValue, crypto:Algorithm.SHA256);
+    any cachedAuthResult = authenticator.getCachedAuthResult(basicAuthCacheKey);
+    if (cachedAuthResult != null) {
+        // cache hit
+        TypeCastError typeCastErr;
+        authInfo, typeCastErr = (AuthenticationInfo) cachedAuthResult;
```
Comment (anupama-pathirage) [inline stdlib/ballerina-builtin/src/main/ballerina/ballerina/security/authentication/basic/authn-interceptor.bal]: Formatting issue : remove space after (AuthenticationInfo)

File: stdlib/ballerina-builtin/src/main/ballerina/ballerina/security/authentication/basic/basic-authenticator.bal [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        return true;
+    }
+    return false;
+}
+
+@Description {value:"Retrieves the cached authentication result if any, for the given basic auth header value"}
+@Param {value:"basicAuthCacheKey: basic authentication cache key - sha256(basic auth header)"}
+@Return {value:"any: cached entry, or null in a cache miss"}
+function <BasicAuthenticator authenticator> getCachedAuthResult (string basicAuthCacheKey) (any) {
+    if (authenticator.authCache != null){
```
Comment (anupama-pathirage) [inline stdlib/ballerina-builtin/src/main/ballerina/ballerina/security/authentication/basic/basic-authenticator.bal]: Formatting issue: Add space after null)

**Comment by isurulucky** [comment]
New PR: https://github.com/ballerina-lang/ballerina/pull/4802

**Pull request closed without merging**